x-pack/filebeat/input/cel/input.go

```diff
@@ -0,0 +1,909 @@
+
+	metrics := newInputMetrics(monitoring.GetNamespace("dataset").GetRegistry(), env.ID)
+	defer metrics.Close()
+
+	ctx := ctxtool.FromCanceller(env.Cancelation)
+
+	client, err := newClient(ctx, cfg, log)
+	if err != nil {
+		return err
+	}
+
+	limiter := newRateLimiterFromConfig(cfg.Resource)
+
+	patterns, err := regexpsFromConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	prg, err := newProgram(cfg.Program, root, client, limiter, patterns)
+	if err != nil {
+		return err
+	}
+
+	var state map[string]interface{}
+	if cfg.State == nil {
+		state = make(map[string]interface{})
+	} else {
+		state = cfg.State
+	}
+	if cursor != nil {
+		state["cursor"] = cursor
+	}
+	goodCursor := cursor
+	goodURL := cfg.Resource.URL.String()
+	state["url"] = goodURL
+	metrics.resource.Set(goodURL)
+	// On entry, state is expected to be in the shape:
+	//
+	// {
+	//     "url": <resource address>,
+	//     "cursor": { ... },
+	//     ...
+	// }
+	//
+	// The url field must be present and can be an HTTP end-point
+	// or a file path. It is currently the responsibility of the
+	// program to handle removing the scheme from a file url if it
+	// is present. The url may be mutated during execution of the
+	// program but the mutated state will not be persisted between
+	// restarts and the url must be present in the returned value
+	// to ensure that it is available in the next evaluation unless
+	// the program has the resource address hard-coded in or it is
+	// available from the cursor.
+	//
+	// Additional fields may be present at the root of the object
+	// and if the program tolerates it, the cursor value may be
+	// absent. Only the cursor is persisted over restarts, but
+	// all fields in state are retained between iterations of
+	// the processing loop except for the produced events array,
+	// see discussion below.
+	//
+	// If the cursor is present the program should perform and
+	// process requests based on its value. If cursor is not
+	// present the program must have alternative logic to
+	// determine what requests to make.
+	//
+	// In addition to this and the functions and globals available
+	// from mito/lib, a global, useragent, is available to use
+	// in requests.
+	err = periodically(ctx, cfg.Interval, func() error {
+		log.Info("process repeated request")
+		var waitUntil time.Time
+		for {
+			// We have a special-case wait for when we have a zero limit.
+			// x/time/rate allow a burst through even when the limit is zero
+			// so in order to ensure that we don't try until we are out of
+			// purgatory we calculate how long we should wait according to
+			// the retry after for a 429 and rate limit headers if we have
+			// a zero rate quota. See handleResponse below.
+			if wait := time.Until(waitUntil); wait > 0 {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(wait):
+				}
+			}
+
+			// Process a set of event requests.
+			log.Debugw("request state", "state", state)
```

Comment:
Not `log := env.Logger.With("cel", cfg.Resource.URL)` at line 104? I don't recall why I had it as "input_url", but that should fix it so that all call on `log` are prefixed with the "cel" namespace, no?

Reply:
The namespace has a different purpose than the name. And what I’m suggesting is not strictly necessary and nor is it a general solution to the problem of using a consistent schema across log messages, but it might help make someone’s job easier.

Basically by adding `logp.Namespace("cel")` it will be creating `cel.state.*` fields in the resulting JSON from the logger as opposed to `state.*`. In case something else is already logging `state`, say as a string, then it will be easier to consume the log data.

Reply:
OK. Is that something that should be added to httpjson as well?